server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java

```diff
@@ -1249,4 +1263,115 @@ public void testPrimaryReceivesDocsDuringReplicaRecovery() throws Exception {
+        final int initialDocCount = scaledRandomIntBetween(100, 200);
+        try (
+            BackgroundIndexer indexer = new BackgroundIndexer(
+                INDEX_NAME,
+                "_doc",
+                client(),
+                -1,
+                RandomizedTest.scaledRandomIntBetween(2, 5),
+                false,
+                random()
+            )
+        ) {
+            indexer.start(initialDocCount);
+            waitForDocs(initialDocCount, indexer);
+            refresh(INDEX_NAME);
+            waitForSearchableDocs(initialDocCount, nodeA, nodeB);
+
+            final int additionalDocCount = scaledRandomIntBetween(0, 200);
+            final int expectedHitCount = initialDocCount + additionalDocCount;
+            indexer.start(additionalDocCount);
+            waitForDocs(expectedHitCount, indexer);
+
+            flushAndRefresh(INDEX_NAME);
+            waitForSearchableDocs(expectedHitCount, nodeA, nodeB);
+
+            ensureGreen(INDEX_NAME);
+            verifyStoreContent();
+        }
+    }
+
+    /**
+     * This test validates that segment replications works for older codec implementations
+     * @throws Exception
+     */
+    public void testWithOlderCodecs() throws Exception {
+        final String nodeA = internalCluster().startNode();
+        final String nodeB = internalCluster().startNode();
+        final Settings settings = Settings.builder()
+            .put(indexSettings())
+            .put(EngineConfig.INDEX_CODEC_SETTING.getKey(), getOlderLuceneCodec())
+            .build();
+        createIndex(INDEX_NAME, settings);
+        final int initialDocCount = scaledRandomIntBetween(100, 200);
+        try (
+            BackgroundIndexer indexer = new BackgroundIndexer(
+                INDEX_NAME,
+                "_doc",
+                client(),
+                -1,
+                RandomizedTest.scaledRandomIntBetween(2, 5),
+                false,
+                random()
+            )
+        ) {
+            indexer.start(initialDocCount);
+            waitForDocs(initialDocCount, indexer);
+            refresh(INDEX_NAME);
+            waitForSearchableDocs(initialDocCount, nodeA, nodeB);
+
+            final int additionalDocCount = scaledRandomIntBetween(0, 200);
+            final int expectedHitCount = initialDocCount + additionalDocCount;
+            indexer.start(additionalDocCount);
+            waitForDocs(expectedHitCount, indexer);
+
+            flushAndRefresh(INDEX_NAME);
+            waitForSearchableDocs(expectedHitCount, nodeA, nodeB);
+
+            ensureGreen(INDEX_NAME);
+            verifyStoreContent();
+        }
+    }
+
+    // Return oldest lucene codec name. It does this by iterating throw all directories under lucene library and filter
+    // ones matching lucene codec name regex
+    private String getOlderLuceneCodec() throws IOException, URISyntaxException {
+        final TreeSet<String> codecNames = new TreeSet<>();
```

Comment:
Names of the lucene codecs can be accessed through Codec.java method availableCodecs(). You can replace codecNames with this and get the oldest codec name to avoid iterating through the directories.

Reply:
Thanks @Poojita-Raj for sharing this. 
`Codec.availableCodecs()` share list of all possible codecs most of which are meant for reading older lucene files and providing backward compatibility. Using these [older codecs](https://github.com/apache/lucene/tree/main/lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs) (older than Lucene90) for IndexWriter (new indices) results in [UnsupportedException](https://github.com/apache/lucene/blob/0c293909c050e2f24ef5a6062d8da31e6595e8cd/lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87StoredFieldsFormat.java#L142) during indexing operation. For new indices, I see only codecs belonging to current major version of 9.x is allowed i.e. Lucene90, Lucene92, Lucene94, Lucene95

Reply:
So this means that this is unsupported for docrep as well right? 
And will the list of lucene jars we get from the lucene jar path you're using always include only the supported codecs even if that list changes in the future?

Reply:
>And will the list of lucene jars we get from the lucene jar path you're using always include only the supported codecs even if that list changes in the future?

Yes, this is how I think lucene organizes the codecs and will continue to do so. It provides separate jar for bwc-codecs for e.g. `org.apache.lucene:lucene-backward-codecs:9.7.0-snapshot-4d1ed9e`. While codecs listed in `org.apache.lucene:lucene-codecs:9.7.0-snapshot-4d1ed9e` provide write compatible codecs (allows indexing). The test uses the later to build codec name list.